### PR TITLE
Move IO right to the edge, allow flexible output and add --verbose flag

### PR DESCRIPTION
**Pull request by sihil**
This replaces the string output type with various output types that can be processed by the UI classes. By allowing this to be returned by the actual methods we grant much greater flexibility in how output is presented (and can be optionally presented using the new `--verbose` flag).

### CODE REVIEW

File: src/main/scala/com/gu/ssm/SSH.scala [hunk 1]
```diff
@@ -102,18 +102,18 @@ object SSH {
...
     val hostsFileString = hostsFile.map(file => s""" -o "UserKnownHostsFile $file" -o "StrictHostKeyChecking yes"""").getOrElse("")
     val connectionString = s"""ssh -o "IdentitiesOnly yes"$useAgentFragment$hostsFileString$targetPortSpecifications -i ${privateKeyFile.getCanonicalFile.toString}${theTTOptions} $user@$ipAddress"""
-    val cmd = if(rawOutput) {
-      s"$connectionString"
-    }else{
-      s"""
-         | # Execute the following command within the next $sshCredentialsLifetimeSeconds seconds:
-         | ${connectionString};
-         |""".stripMargin
+    val cmd = if (rawOutput) {
```
Comment (LATaylor-guardian) [inline src/main/scala/com/gu/ssm/SSH.scala]: This pattern is used in three places within this file - maybe you could pull out into a small function?

File: src/main/scala/com/gu/ssm/SSH.scala [hunk 2]
```diff
@@ -102,18 +102,18 @@ object SSH {
...
+      Seq(
+        Metadata(s"# Execute the following command within the next $sshCredentialsLifetimeSeconds seconds:"),
+        Out(s"$connectionString;")
+      )
     }
     (instance.id, cmd)
   }
 
-  def sshCmdBastion(rawOutput: Boolean)(privateKeyFile: File, bastionInstance: Instance, targetInstance: Instance, targetInstanceUser: String, bastionIpAddress: String, targetIpAddress: String, bastionPortNumberOpt: Option[Int], bastionUser: String, targetInstancePortNumberOpt: Option[Int], useAgent: Option[Boolean], hostsFile: Option[File]): (InstanceId, String) = {
+  def sshCmdBastion(rawOutput: Boolean)(privateKeyFile: File, bastionInstance: Instance, targetInstance: Instance, targetInstanceUser: String, bastionIpAddress: String, targetIpAddress: String, bastionPortNumberOpt: Option[Int], bastionUser: String, targetInstancePortNumberOpt: Option[Int], useAgent: Option[Boolean], hostsFile: Option[File]): (InstanceId, Seq[Output]) = {
```
Comment (LATaylor-guardian) [inline src/main/scala/com/gu/ssm/SSH.scala]: Would be great to have some type safety around the String arguments to this function. Outside of this PR though.

File: src/main/scala/com/gu/ssm/SSH.scala [hunk 3]
```diff
@@ -127,20 +127,20 @@ object SSH {
...
+      )
     }
     (targetInstance.id, cmd)
   }
 
   // The first file goes to the second file
   // The remote file is indicated by a colon
 
-  def scpCmdStandard(rawOutput: Boolean)(privateKeyFile: File, instance: Instance, user: String, ipAddress: String, targetInstancePortNumberOpt: Option[Int], useAgent: Option[Boolean], hostsFile: Option[File], sourceFile: String, targetFile: String): (InstanceId, String) = {
+  def scpCmdStandard(rawOutput: Boolean)(privateKeyFile: File, instance: Instance, user: String, ipAddress: String, targetInstancePortNumberOpt: Option[Int], useAgent: Option[Boolean], hostsFile: Option[File], sourceFile: String, targetFile: String): (InstanceId, Seq[Output]) = {
```
Comment (LATaylor-guardian) [inline src/main/scala/com/gu/ssm/SSH.scala]: As above.

File: src/main/scala/com/gu/ssm/SSH.scala (hunk 1 above)
Comment (shtukas) [reply src/main/scala/com/gu/ssm/SSH.scala]: Good point

**Comment by LATaylor-guardian** [comment]
👍

**Pull request merged**